### PR TITLE
[Protocol 3.6] Protocol 3.6 improvements

### PR DESCRIPTION
**Pull request by Brechtpd**
- #1462 
- #1457 
- #1456 
- #1422
- #1209 
- #1279 (only things in the protocol)
- Removed index
- Updated to solidity 0.7.0
- Removed new account/transfer account transaction types

And some other small improvements.

### CODE REVIEW

**Comment by dong77** [comment]
Brecht, when you add the new field to an Order, can you make sure it means the other order's id: otherOrderID & newField == newField, instead of checking orderOrderID == newField. Or do you think it's a good idea?

**Comment by Brechtpd** [comment]
> Brecht, when you add the new field to an Order, can you make sure it means the other order's id: otherOrderID & newField == newField, instead of checking orderOrderID == newField. Or do you think it's a good idea?

Not sure I follow, why check the orderID? To lock an order to a single other order? But the owner of the order could still make whatever order he wants in that case because he can create as many orders with that orderID as he wants, so I don't think that could work. I guess we could check the other's order hash, but I'm not sure what you're trying to achieve here. :)

**Comment by dong77** [comment]
My bad, wrong expression, I did mean the owner's address of the other order.

> > Brecht, when you add the new field to an Order, can you make sure it means the other order's id: otherOrderID & newField == newField, instead of checking orderOrderID == newField. Or do you think it's a good idea?
> 
> Not sure I follow, why check the orderID? To lock an order to a single other order? But the owner of the order could still make whatever order he wants in that case because he can create as many orders with that orderID as he wants, so I don't think that could work. I guess we could check the other's order hash, but I'm not sure what you're trying to achieve here. :)

My bad, wrong expression. Let me try again.
Assume the new field is called `addressMask`, now what does this means：it means the other order's owner address `order.owner` must satisfy: 
```
ower.owner & addressMask == addressMask
```

instead of：
```
ower.owner == addressMask
```

The benefit is that we can allow the order to be matched by one of N owners in a set, we just need to do:
```
addressMask = owner1 | owner2 | .. |ownerN
```

This is more flexible, but I'm not sure it will be used a lot.

**Comment by Brechtpd** [comment]
Ah, some sort of [bloom filter](https://en.wikipedia.org/wiki/Bloom_filter). Yeah, could be cool and is still easy to implement. Not sure how "safe" it is when the set is quite large, but for a set of a single element it will be as safe as the equality check, so can't seem to hurt.

**Comment by Brechtpd** [comment]
To do bitwise operations we need to break the value up in bits and then do the actual bit operation. So this will cost around 160 * 4 * 2 = 1280 constraints. So not cheap, but also not expensive.

`order.owner & addressMask == addressMask` needs to be `order.owner & addressMask == owner` I think. 

Because the set contains ORd addresses I think the addressMask will contain a lot of 1s even with a few addresses(~75% 1s for the result of ORing 2 addresses?), so I think it won't be that hard to brute force find addresses that will pass the test that were not in the original set if the set is > 1 or 2 (but it does have the nice property that only gets worse with the size of the set).

**Comment by Brechtpd** [comment]
I did some testing with the address mask, and the false positives are simple too high even for extremely small sets. I used this python script to verify:
```
import random

numRounds = 100000
setSize = 4

mask = 0
set = []
for i in range(setSize):
    value = random.SystemRandom().randint(0, 2**160)
    mask |= value
    set.append(value)
print("mask: " + "{0:b}".format(mask))

numFalsePositives = 0
breakingPoint = 0
found = 0
for i in range(numRounds):
    address = random.SystemRandom().randint(0, 2**160)
    if address & mask == address:
        if not (address in set):
            numFalsePositives += 1

print("False positive rate: " + str(numFalsePositives*100/numRounds) + "%")
```

It takes a fraction of a second to find a false positive even with just 4 values in the set. And even a set of 1 has quite a lot of false positives (these are just addresses that have some of the bits not set to 1, which there are also many of).

Making it an actual bloom filter isn't trivial and will also cost a lot more constraints.

Anther option is to just have a fixed, limited set of addresses (e.g. max 16 addresses) and just hash those to create a mask, the operator than needs to show that the target address is part of this fixed list of addresses, but this also quite a bit of work to implement.

So would stick to the simple equality check unless a simple alternative is found which at least works and is 100% secure for the common case it just needs to target a single address. In any case the number of addresses in the set will be limited no matter the method (unless with great cost). A limited set of target addresses can also be made possible with order aliasing: the exact same order (with the same orderID) is simply signed multiple times with different taker addresses.

**Comment by dong77** [comment]
> I did some testing with the address mask, and the false positives are simple too high even for extremely small sets. I used this python script to verify:
> 
> ```
> import random
> 
> numRounds = 100000
> setSize = 4
> 
> mask = 0
> set = []
> for i in range(setSize):
>     value = random.SystemRandom().randint(0, 2**160)
>     mask |= value
>     set.append(value)
> print("mask: " + "{0:b}".format(mask))
> 
> numFalsePositives = 0
> found = 0
> for i in range(numRounds):
>     address = random.SystemRandom().randint(0, 2**160)
>     if address & mask == address:
>         if not (address in set):
>             numFalsePositives += 1
> 
> print("False positive rate: " + str(numFalsePositives*100/numRounds) + "%")
> ```
> 
> It takes a fraction of a second to find a false positive even with just 4 values in the set. And even a set of 1 has quite a lot of false positives (these are just addresses that have some of the bits not set to 1, which there are also many of). The problem is that the mask just contains too many 1s even for very small sets. The chance a bit is 1 is ` 1 - (1/2)^n`, which quickly goes to 1.
> 
> Making it an actual bloom filter isn't trivial and will also cost a lot more constraints.
> 
> Anther option is to just have a fixed, limited set of addresses (e.g. max 16 addresses) and just hash those to create a hash of the set, the operator than needs to show that the target address is part of this fixed list of addresses, but this also quite a bit of work to implement and doesn't really add that much.
> 
> A limited set of target addresses can also be made possible with order aliasing: the exact same order (with the same orderID) is simply signed multiple times with different taker addresses. I think this is currently the best method.
> 
> So would stick to the simple equality check unless a simple alternative is found which at works for decent sized sets and is 100% secure for the common case it just needs to target a single address. In any case the number of addresses in the set will be limited no matter the method (unless with great cost).

So lets keep things simple by checking equality then.

File: packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol [hunk 1]
```diff
@@ -142,6 +128,23 @@ library OwnerChangeTransaction
             // If the account has a wallet, use it to recover the account
             require(accountTransfer.walletHash != 0, "ACCOUNT_HAS_NO_WALLET");
 
+            // Verify authorization from the new owner
```
Comment (dong77) [inline packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol:131]: I don't think we need to get authorization from the new owner, otherwise, the user experience will be bad. Even in hebao, we don't need a signature from the new owner.

File: packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol [hunk 2]
```diff
@@ -153,6 +156,14 @@ library OwnerChangeTransaction
                     )
                 )
             );
+            // Verify that the data was signed by the current owner
```
Comment (dong77) [inline packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol:159]: In the case of using stateless wallet, the current owner may have lost access, so no need to check authorization from the current owner.

File: packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol [hunk 3]
```diff
@@ -142,6 +128,23 @@ library OwnerChangeTransaction
             // If the account has a wallet, use it to recover the account
             require(accountTransfer.walletHash != 0, "ACCOUNT_HAS_NO_WALLET");
 
+            // Verify authorization from the new owner
+            if (auxData.signature.length > 0) {
+                require(
```
Comment (dong77) [inline packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol:133]: Across many files, we find this pattern of verifying either ECDSA signature or an approval hash. Maybe we can extract a common function?

File: packages/loopring_v3.js/src/constants.ts [hunk 4]
```diff
@@ -3,19 +3,18 @@ import { FloatEncoding } from "./float";
...
 
 export class Constants {
-  static readonly BINARY_TREE_DEPTH_TRADING_HISTORY = 14;
-  static readonly BINARY_TREE_DEPTH_ACCOUNTS = 24;
+  static readonly BINARY_TREE_DEPTH_STORAGE = 14;
+  static readonly BINARY_TREE_DEPTH_ACCOUNTS = 32;
   static readonly BINARY_TREE_DEPTH_TOKENS = 12;
 
-  static readonly TX_DATA_AVAILABILITY_SIZE = 104;
+  static readonly TX_DATA_AVAILABILITY_SIZE = 106;
```
[the remark this answers is not in the document]
Comment (Brechtpd) [reply packages/loopring_v3.js/src/constants.ts:10]: Yes, no way around that, we have to hash the same amount of data for each transaction type. The zeros are compressed away where it makes sense (though the compression can certainly still be approved).

File: packages/loopring_v3/contracts/impl/BasicDepositContract.sol [hunk 5]
```diff
@@ -115,15 +116,15 @@ contract BasicDepositContract is IDepositContract, ReentrancyGuard, Claimable
         override
         payable
         onlyExchange
-        nonReentrant
+        //nonReentrant
```
[the remark this answers is not in the document]
Comment (Brechtpd) [reply packages/loopring_v3/contracts/impl/BasicDepositContract.sol:119]: Never, because only the exchange can call these functions and the exchange is always nonReentrant. Basically the deposit contract is like an upgradeable library. Removing this saves ~3,000 gas per call into the deposit contract.

File: packages/loopring_v3/contracts/impl/agents/FastWithdrawalAgent.sol [hunk 6]
```diff
@@ -100,34 +99,59 @@ contract FastWithdrawalAgent is ReentrancyGuard
...
+                        fastWithdrawal.validUntil
+                    )
+                )
+            );
+
+            // Check the signature
+            require(hash.verifySignature(fastWithdrawal.from, fastWithdrawal.signature), "INVALID_SIGNATURE");
+
+            // Check the time limit
+            require(block.timestamp <= fastWithdrawal.validUntil, "TX_EXPIRED");
```
[the remark this answers is not in the document]
Comment (Brechtpd) [reply packages/loopring_v3/contracts/impl/agents/FastWithdrawalAgent.sol:129]: Yeah, the` validUntil` here isn't passed into `setWithdrawalRecipient`, and the user never signs any of the data used in that case so basically at this point any data can be used, of course as the risk of the liquidity provider.

File: packages/loopring_v3/contracts/impl/agents/OpenGSN2Agent.sol [hunk 7]
```diff
@@ -1,169 +1,169 @@
...
-    {
-        (bool success, bytes memory returnData) =
-            exchange.call{value: msg.value}(msg.data);
-
-        if (!success) {
-            assembly { revert(add(returnData, 32), mload(returnData)) }
-        }
-    }
-}
+// import "@opengsn/gsn/contracts/BaseRelayRecipient.sol";
```
[the remark this answers is not in the document]
Comment (Brechtpd) [reply packages/loopring_v3/contracts/impl/agents/OpenGSN2Agent.sol:6]: Doesn't compile yet with 0.7.0, so uncommented until it gets upgraded I'm afraid...

File: packages/loopring_v3/contracts/impl/libtransactions/WithdrawTransaction.sol [hunk 8]
```diff
@@ -170,6 +170,23 @@ library WithdrawTransaction
...
+        // Check if there is a withdrawal recipient
+        address recipient = S.withdrawalRecipient[withdrawal.owner][withdrawal.to][withdrawal.tokenID][withdrawal.amount][withdrawal.nonce];
+        if (recipient != address(0)) {
+            // Auxiliary data is not supported
+            require (auxData.auxiliaryData.length == 0, "AUXILIARY_DATA_NOT_ALLOWED");
+
+            // Set the new recipient address
+            withdrawal.to = recipient;
+            // Allow any amount of gas to be used on this withdrawal (which allows the transfer to be skipped)
+            withdrawal.minGas = 0;
```
[the remark this answers is not in the document]
Comment (Brechtpd) [reply packages/loopring_v3/contracts/impl/libtransactions/WithdrawTransaction.sol:182]: Otherwise your idea won't work to do cheap redirects without actual token transfers. If the user does a normal withdrawal then surely in most cases some minimum amount of gas will be specified. So if we still keep enforcing this gas limit then the operator can't choose to use the cheaper path of just updating `amountWithdrawable` multiple times for the liquidity provider and he would still be forced to do the (likely) more expensive full token transfer withdrawals.

File: packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol (hunk 1 above)
Comment (Brechtpd) [reply packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol:131]: I removed the requirement when the old owner can still sign, but kept this one because someone needs to sign something. Basically we have the same problem as with our smart wallet: who can sign meta txs for an account. The new owner is required to sign because he authorizes:
- The fee payment (using funds from the account being recovered, so only the old or the new owner can do that)
- Which data is used to recover the account

I guess we could let another account pay for the recovery, but I think that makes things a bit more complicated, and in any case someone needs to sign the transaction so why not the new owner and keep things simple?

File: packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol (hunk 2 above)
Comment (Brechtpd) [reply packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol:159]: The wallet data would be signed before it is set on the account while the owner still has access to his keys. So if the owner wants to set the data on his account, he would also sign it and store the signature for it.

I added this for extra security when the user's EdDSA keys are lost or shared between different users. Currently it is allowed to update the EdDSA keys/wallet data for an account with just the EdDSA keys because it's significantly cheaper than needing an ECDSA signature. If the users's EdDSA keys are stolen all funds in his account are at risk, but when the wallet can also be updated he would also risk losing his account completely because the owner could be changed by setting some dummy wallet. By also requiring an ECDSA signature for the wallet data this is not possible. At least the user's account is safe from being overtaken completely, so the user's accountID is still secure as long as his Ethereum key is safe.

So not too important, but allows us to keep more common operations as cheap as possible while having some extra safe guards against theft of the EdDSA keys.

File: packages/loopring_v3/contracts/impl/libexchange/ExchangeWithdrawals.sol [hunk 9]
```diff
@@ -126,23 +126,22 @@ library ExchangeWithdrawals
...
         external
     {
         uint16 tokenID = S.getTokenID(token);
-        ExchangeData.Deposit storage deposit = S.pendingDeposits[owner][tokenID][index];
+        ExchangeData.Deposit storage deposit = S.pendingDeposits[owner][tokenID];
         require(deposit.timestamp != 0, "DEPOSIT_NOT_WITHDRAWABLE_YET");
 
         // Check if the deposit has indeed exceeded the time limit
-        require(now >= deposit.timestamp + S.maxAgeDepositUntilWithdrawable, "DEPOSIT_NOT_WITHDRAWABLE_YET");
+        require(block.timestamp >= deposit.timestamp + S.maxAgeDepositUntilWithdrawable, "DEPOSIT_NOT_WITHDRAWABLE_YET");
```
Comment (dong77) [inline packages/loopring_v3/contracts/impl/libexchange/ExchangeWithdrawals.sol:138]: should we skip check this if the exchange is in WithdrawalMode? something like:

```solidity
require(isWithdrawalMode ||
             block.timestamp >= deposit.timestamp + S.maxAgeDepositUntilWithdrawable, "DEPOSIT_NOT_WITHDRAWABLE_YET");
```

File: packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol (hunk 2 above)
Comment (dong77) [reply packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol:159]: "The wallet data would be signed before it is set on the account while the owner still has access to his keys"
Unless the signature of the current owner is saved inside the Merkle tree and becomes part of DA, there is no guarantee that his/her signature can be found when the user's wallet is used.

File: packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol (hunk 1 above)
Comment (dong77) [reply packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol:131]: We can/should always assume that if the current owner configs his wallet to have guardians, then a majority of the guardian's approval does authorize the operator to charge the recovery fee out of this account. I don't think the new owner needs to sign. Most likely one of his guardians is another address of his.

"Which data is used to recover the account"
I think we can simply reset the wallet data after the recovery, let the new owner to config his wallet later.

**Comment by dong77** [comment]
Please see - https://github.com/Loopring/protocols/issues/1523

File: packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol (hunk 2 above)
Comment (Brechtpd) [reply packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol:159]: The same is true for the actual wallet data, which is also not part of the Merkle tree or the DA, but needs to be stored somewhere else by the user/platform (publicly or privately), so no difference there.

File: packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol (hunk 1 above)
Comment (Brechtpd) [reply packages/loopring_v3/contracts/impl/libtransactions/AccountTransferTransaction.sol:131]: That's a good point, but they'd have approve the fee payment as well which they currently do not, but that makes sense.
